test/model/typing_status_test.dart

```diff
@@ -0,0 +1,223 @@
+import 'package:checks/checks.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/api/model/events.dart';
+import 'package:zulip/api/model/model.dart';
+import 'package:zulip/model/narrow.dart';
+import 'package:zulip/model/typing_status.dart';
+
+import '../example_data.dart' as eg;
+import '../fake_async.dart';
+
+void main() {
+  late TypingStatus model;
+  late int notifiedCount;
+
+  void checkNotNotified() {
+    check(notifiedCount).equals(0);
+  }
+
+  void checkNotifiedOnce() {
+    check(notifiedCount).equals(1);
+    notifiedCount = 0;
+  }
+
+  void handleTypingEvent(SendableNarrow narrow, TypingOp op, User sender) {
+    assert(sender != eg.selfUser);
+    model.handleTypingEvent(eg.typingEvent(narrow, op, sender.userId));
+  }
+
+  void checkTypists(Map<SendableNarrow, List<User>> typistsByNarrow) {
+    final actualTypistsByNarrow = <SendableNarrow, Iterable<int>>{};
+    for (final narrow in model.debugActiveNarrows) {
+      actualTypistsByNarrow[narrow] = model.typistIdsInNarrow(narrow);
+    }
+    check(actualTypistsByNarrow).deepEquals(
+      typistsByNarrow.map((k, v) => MapEntry(k, v.map((e) => e.userId))));
+  }
+
+  void prepareModel({
+    int? selfUserId,
+    Map<SendableNarrow, List<User>> typistsByNarrow = const {},
+  }) {
+    model = TypingStatus(
+      selfUserId: selfUserId ?? eg.selfUser.userId,
+      typingStartedExpiryPeriod: const Duration(milliseconds: 15000));
+    check(model.debugActiveNarrows).isEmpty();
+    notifiedCount = 0;
+    model.addListener(() => notifiedCount += 1);
+
+    typistsByNarrow.forEach((narrow, typists) {
+      for (final typist in typists) {
+        handleTypingEvent(narrow, TypingOp.start, typist);
+        checkNotifiedOnce();
+      }
+    });
+    checkTypists(typistsByNarrow);
+  }
+
+  final stream = eg.stream();
+  final topicNarrow = TopicNarrow(stream.streamId, 'foo');
+
+  final dmNarrow = DmNarrow.withUser(eg.otherUser.userId, selfUserId: eg.selfUser.userId);
+  final groupNarrow = DmNarrow.withOtherUsers(
+    [eg.otherUser.userId, eg.thirdUser.userId], selfUserId: eg.selfUser.userId);
+
+  group('handle typing start events', () {
```

Comment:
Another case to have an explicit test case for: a start event in a topic narrow.

(This gets covered by the "remove typists from different narrows" case; but because it's not the point of the test, it'd be easy for that to go away in a future revision of the tests and then that coverage would get dropped.)